Fails with "java.lang.StackOverflowError"

`gradle generateDependencyGraph --stacktrace` only generates a StackOverflowError for me using Version 0.4.0 of the plugin:

```
Caused by: java.lang.StackOverflowError
        at kotlin.text.StringsKt__StringsKt.findAnyOf$StringsKt__StringsKt(Strings.kt:872)
        at kotlin.text.StringsKt__StringsKt.access$findAnyOf(Strings.kt:1)
        at kotlin.text.StringsKt__StringsKt$rangesDelimitedBy$4.invoke(Strings.kt:1141)
        at kotlin.text.StringsKt__StringsKt$rangesDelimitedBy$4.invoke(Strings.kt)
        at kotlin.text.DelimitedRangesSequence$iterator$1.calcNext(Strings.kt:1069)
        at kotlin.text.DelimitedRangesSequence$iterator$1.hasNext(Strings.kt:1099)
        at kotlin.sequences.TransformingSequence$iterator$1.hasNext(Sequences.kt:153)
        at kotlin.sequences.SequencesKt___SequencesKt.joinTo(_Sequences.kt:1657)
        at kotlin.sequences.SequencesKt___SequencesKt.joinToString(_Sequences.kt:1677)
        at kotlin.sequences.SequencesKt___SequencesKt.joinToString$default(_Sequences.kt:1676)
        at kotlin.text.StringsKt__StringsJVMKt.replace(StringsJVM.kt:66)
        at kotlin.text.StringsKt__StringsJVMKt.replace$default(StringsJVM.kt:65)
        at com.vanniktech.dependency.graph.generator.ExtensionsKt.getDotIdentifier(extensions.kt:5)
        at com.vanniktech.dependency.graph.generator.DotGenerator.append(DotGenerator.kt:57)
        at com.vanniktech.dependency.graph.generator.DotGenerator.append(DotGenerator.kt:71)
        at com.vanniktech.dependency.graph.generator.DotGenerator.append(DotGenerator.kt:71)
        at com.vanniktech.dependency.graph.generator.DotGenerator.append(DotGenerator.kt:71)
...
```

Unfortunately I can't share details of the project besides that it is pretty huge :wink:

Any hints what might cause it to endless loop in DotGenerator.kt:71?

Same problem here. I've tried also with v0.3.0 and got the same:

`FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':generateDependencyGraph'.
> java.lang.StackOverflowError (no error message)

* Try:
Run with --info or --debug option to get more log output.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':generateDependencyGraph'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:100)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:70)
        at org.gradle.api.internal.tasks.execution.OutputDirectoryCreatingTaskExecuter.execute(OutputDirectoryCreatingTaskExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:60)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:97)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:87)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:52)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:248)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:336)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:328)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:199)
        at org.gradle.internal.progress.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:110)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:241)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:230)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.processTask(DefaultTaskPlanExecutor.java:123)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.access$200(DefaultTaskPlanExecutor.java:79)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:104)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:98)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:626)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:581)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:98)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: java.lang.StackOverflowError
        at kotlin.collections.CollectionsKt___CollectionsKt.single(_Collections.kt:471)
        at kotlin.collections.CollectionsKt___CollectionsKt.single(_Collections.kt:454)
        at kotlin.text.StringsKt__StringsKt.findAnyOf$StringsKt__StringsKt(Strings.kt:854)
`

--- a/src/main/kotlin/com/vanniktech/dependency/graph/generator/DotGenerator.kt
+++ b/src/main/kotlin/com/vanniktech/dependency/graph/generator/DotGenerator.kt
@@ -53,19 +53,17 @@ internal class DotGenerator(
   }
 
   private fun append(dependency: ResolvedDependency, parentIdentifier: String, content: StringBuilder) {
-    if (generator.include.invoke(dependency)) {
-      val identifier = (dependency.moduleGroup + dependency.moduleName).dotIdentifier
+    val identifier = (dependency.moduleGroup + dependency.moduleName).dotIdentifier
+    val pair = parentIdentifier to identifier
 
-      val pair = parentIdentifier to identifier
-      if (!addedConnections.contains(pair)) { // We don't want to re-add the same dependencies.
-        addedConnections.add(pair)
+    if (generator.include.invoke(dependency) && !addedConnections.contains(pair)) {
+      addedConnections.add(pair)
 
-        if (!addedDependencies.contains(identifier)) {
-          content.append("  $identifier ${generator.dependencyFormattingOptions.invoke(dependency).withLabel(dependency.getDisplayName())};\n")
-        }
-
-        content.append("  $parentIdentifier -> $identifier;\n")
+      if (!addedDependencies.contains(identifier)) {
+        content.append("  $identifier ${generator.dependencyFormattingOptions.invoke(dependency).withLabel(dependency.getDisplayName())};\n")
       }
+
+      content.append("  $parentIdentifier -> $identifier;\n")
 
       if (generator.children.invoke(dependency)) {
         dependency.children.forEach { append(it, identifier, content) }

--- a/src/test/java/com/vanniktech/dependency/graph/generator/DotGeneratorTest.kt
+++ b/src/test/java/com/vanniktech/dependency/graph/generator/DotGeneratorTest.kt
@@ -181,6 +181,140 @@ class DotGeneratorTest {
         |""".trimMargin())
   }
 
+  @Test fun recursiveDependencies() {
+      singleEmpty.dependencies.add("implementation", "org.apache.xmlgraphics:batik-gvt:1.7")
+
+      assertThat(DotGenerator(singleEmpty, ALL).generateContent()).isEqualTo("""
+        |digraph G {
+        |  singleempty [label="singleempty", shape="box"];
+        |  orgapachexmlgraphicsbatikgvt [label="batik-gvt", shape="box"];
+        |  singleempty -> orgapachexmlgraphicsbatikgvt;
+        |  orgapachexmlgraphicsbatikbridge [label="batik-bridge", shape="box"];
+        |  orgapachexmlgraphicsbatikgvt -> orgapachexmlgraphicsbatikbridge;
+        |  orgapachexmlgraphicsbatikscript [label="batik-script", shape="box"];
+        |  orgapachexmlgraphicsbatikbridge -> orgapachexmlgraphicsbatikscript;
+        |  orgapachexmlgraphicsbatikbridge [label="batik-bridge", shape="box"];
+        |  orgapachexmlgraphicsbatikscript -> orgapachexmlgraphicsbatikbridge;
+        |  orgapachexmlgraphicsbatikgvt [label="batik-gvt", shape="box"];
+        |  orgapachexmlgraphicsbatikbridge -> orgapachexmlgraphicsbatikgvt;
+        |  orgapachexmlgraphicsbatikawtutil [label="batik-awt-util", shape="box"];
+        |  orgapachexmlgraphicsbatikgvt -> orgapachexmlgraphicsbatikawtutil;
+        |  orgapachexmlgraphicsbatikutil [label="batik-util", shape="box"];
+        |  orgapachexmlgraphicsbatikawtutil -> orgapachexmlgraphicsbatikutil;
+        |  orgapachexmlgraphicsbatikutil [label="batik-util", shape="box"];
+        |  orgapachexmlgraphicsbatikgvt -> orgapachexmlgraphicsbatikutil;
+        |  xmlapisxmlapis [label="xml-apis", shape="box"];
+        |  orgapachexmlgraphicsbatikgvt -> xmlapisxmlapis;
+        |  orgapachexmlgraphicsbatiksvgdom [label="batik-svg-dom", shape="box"];
+        |  orgapachexmlgraphicsbatikbridge -> orgapachexmlgraphicsbatiksvgdom;
+        |  orgapachexmlgraphicsbatikanim [label="batik-anim", shape="box"];
+        |  orgapachexmlgraphicsbatiksvgdom -> orgapachexmlgraphicsbatikanim;
+        |  orgapachexmlgraphicsbatiksvgdom [label="batik-svg-dom", shape="box"];
+        |  orgapachexmlgraphicsbatikanim -> orgapachexmlgraphicsbatiksvgdom;
+        |  orgapachexmlgraphicsbatikparser [label="batik-parser", shape="box"];
+        |  orgapachexmlgraphicsbatiksvgdom -> orgapachexmlgraphicsbatikparser;
+        |  orgapachexmlgraphicsbatikawtutil [label="batik-awt-util", shape="box"];
+        |  orgapachexmlgraphicsbatikparser -> orgapachexmlgraphicsbatikawtutil;
+        |  orgapachexmlgraphicsbatikxml [label="batik-xml", shape="box"];
+        |  orgapachexmlgraphicsbatikparser -> orgapachexmlgraphicsbatikxml;
+        |  orgapachexmlgraphicsbatikutil [label="batik-util", shape="box"];
+        |  orgapachexmlgraphicsbatikxml -> orgapachexmlgraphicsbatikutil;
+        |  orgapachexmlgraphicsbatikutil [label="batik-util", shape="box"];
+        |  orgapachexmlgraphicsbatikparser -> orgapachexmlgraphicsbatikutil;
+        |  orgapachexmlgraphicsbatikawtutil [label="batik-awt-util", shape="box"];
+        |  orgapachexmlgraphicsbatiksvgdom -> orgapachexmlgraphicsbatikawtutil;
+        |  orgapachexmlgraphicsbatikdom [label="batik-dom", shape="box"];
+        |  orgapachexmlgraphicsbatiksvgdom -> orgapachexmlgraphicsbatikdom;
+        |  orgapachexmlgraphicsbatikcss [label="batik-css", shape="box"];
+        |  orgapachexmlgraphicsbatikdom -> orgapachexmlgraphicsbatikcss;
+        |  orgapachexmlgraphicsbatikutil [label="batik-util", shape="box"];
+        |  orgapachexmlgraphicsbatikcss -> orgapachexmlgraphicsbatikutil;
+        |  orgapachexmlgraphicsbatikext [label="batik-ext", shape="box"];
+        |  orgapachexmlgraphicsbatikcss -> orgapachexmlgraphicsbatikext;
+        |  xmlapisxmlapis [label="xml-apis", shape="box"];
+        |  orgapachexmlgraphicsbatikext -> xmlapisxmlapis;
+        |  xmlapisxmlapis [label="xml-apis", shape="box"];
+        |  orgapachexmlgraphicsbatikcss -> xmlapisxmlapis;
+        |  xmlapisxmlapisext [label="xml-apis-ext", shape="box"];
+        |  orgapachexmlgraphicsbatikcss -> xmlapisxmlapisext;
+        |  orgapachexmlgraphicsbatikxml [label="batik-xml", shape="box"];
+        |  orgapachexmlgraphicsbatikdom -> orgapachexmlgraphicsbatikxml;
+        |  orgapachexmlgraphicsbatikutil [label="batik-util", shape="box"];
+        |  orgapachexmlgraphicsbatikdom -> orgapachexmlgraphicsbatikutil;
+        |  orgapachexmlgraphicsbatikext [label="batik-ext", shape="box"];
+        |  orgapachexmlgraphicsbatikdom -> orgapachexmlgraphicsbatikext;
+        |  xalanxalan [label="xalan", shape="box"];
+        |  orgapachexmlgraphicsbatikdom -> xalanxalan;
+        |  xmlapisxmlapis [label="xml-apis", shape="box"];
+        |  xalanxalan -> xmlapisxmlapis;
+        |  xmlapisxmlapis [label="xml-apis", shape="box"];
+        |  orgapachexmlgraphicsbatikdom -> xmlapisxmlapis;
+        |  xmlapisxmlapisext [label="xml-apis-ext", shape="box"];
+        |  orgapachexmlgraphicsbatikdom -> xmlapisxmlapisext;
+        |  orgapachexmlgraphicsbatikcss [label="batik-css", shape="box"];
+        |  orgapachexmlgraphicsbatiksvgdom -> orgapachexmlgraphicsbatikcss;
+        |  orgapachexmlgraphicsbatikutil [label="batik-util", shape="box"];
+        |  orgapachexmlgraphicsbatiksvgdom -> orgapachexmlgraphicsbatikutil;
+        |  orgapachexmlgraphicsbatikext [label="batik-ext", shape="box"];
+        |  orgapachexmlgraphicsbatiksvgdom -> orgapachexmlgraphicsbatikext;
+        |  xmlapisxmlapis [label="xml-apis", shape="box"];
+        |  orgapachexmlgraphicsbatiksvgdom -> xmlapisxmlapis;
+        |  xmlapisxmlapisext [label="xml-apis-ext", shape="box"];
+        |  orgapachexmlgraphicsbatiksvgdom -> xmlapisxmlapisext;
+        |  orgapachexmlgraphicsbatikparser [label="batik-parser", shape="box"];
+        |  orgapachexmlgraphicsbatikanim -> orgapachexmlgraphicsbatikparser;
+        |  orgapachexmlgraphicsbatikawtutil [label="batik-awt-util", shape="box"];
+        |  orgapachexmlgraphicsbatikanim -> orgapachexmlgraphicsbatikawtutil;
+        |  orgapachexmlgraphicsbatikdom [label="batik-dom", shape="box"];
+        |  orgapachexmlgraphicsbatikanim -> orgapachexmlgraphicsbatikdom;
+        |  orgapachexmlgraphicsbatikutil [label="batik-util", shape="box"];
+        |  orgapachexmlgraphicsbatikanim -> orgapachexmlgraphicsbatikutil;
+        |  orgapachexmlgraphicsbatikext [label="batik-ext", shape="box"];
+        |  orgapachexmlgraphicsbatikanim -> orgapachexmlgraphicsbatikext;
+        |  xmlapisxmlapis [label="xml-apis", shape="box"];
+        |  orgapachexmlgraphicsbatikanim -> xmlapisxmlapis;
+        |  xmlapisxmlapisext [label="xml-apis-ext", shape="box"];
+        |  orgapachexmlgraphicsbatikanim -> xmlapisxmlapisext;
+        |  orgapachexmlgraphicsbatikanim [label="batik-anim", shape="box"];
+        |  orgapachexmlgraphicsbatikbridge -> orgapachexmlgraphicsbatikanim;
+        |  orgapachexmlgraphicsbatikparser [label="batik-parser", shape="box"];
+        |  orgapachexmlgraphicsbatikbridge -> orgapachexmlgraphicsbatikparser;
+        |  orgapachexmlgraphicsbatikawtutil [label="batik-awt-util", shape="box"];
+        |  orgapachexmlgraphicsbatikbridge -> orgapachexmlgraphicsbatikawtutil;
+        |  orgapachexmlgraphicsbatikdom [label="batik-dom", shape="box"];
+        |  orgapachexmlgraphicsbatikbridge -> orgapachexmlgraphicsbatikdom;
+        |  orgapachexmlgraphicsbatikcss [label="batik-css", shape="box"];
+        |  orgapachexmlgraphicsbatikbridge -> orgapachexmlgraphicsbatikcss;
+        |  orgapachexmlgraphicsbatikxml [label="batik-xml", shape="box"];
+        |  orgapachexmlgraphicsbatikbridge -> orgapachexmlgraphicsbatikxml;
+        |  orgapachexmlgraphicsbatikutil [label="batik-util", shape="box"];
+        |  orgapachexmlgraphicsbatikbridge -> orgapachexmlgraphicsbatikutil;
+        |  orgapachexmlgraphicsbatikext [label="batik-ext", shape="box"];
+        |  orgapachexmlgraphicsbatikbridge -> orgapachexmlgraphicsbatikext;
+        |  xalanxalan [label="xalan", shape="box"];
+        |  orgapachexmlgraphicsbatikbridge -> xalanxalan;
+        |  xmlapisxmlapis [label="xml-apis", shape="box"];
+        |  orgapachexmlgraphicsbatikbridge -> xmlapisxmlapis;
+        |  xmlapisxmlapisext [label="xml-apis-ext", shape="box"];
+        |  orgapachexmlgraphicsbatikbridge -> xmlapisxmlapisext;
+        |  orgapachexmlgraphicsbatiksvgdom [label="batik-svg-dom", shape="box"];
+        |  orgapachexmlgraphicsbatikscript -> orgapachexmlgraphicsbatiksvgdom;
+        |  orgapachexmlgraphicsbatikdom [label="batik-dom", shape="box"];
+        |  orgapachexmlgraphicsbatikscript -> orgapachexmlgraphicsbatikdom;
+        |  orgapachexmlgraphicsbatikutil [label="batik-util", shape="box"];
+        |  orgapachexmlgraphicsbatikscript -> orgapachexmlgraphicsbatikutil;
+        |  orgapachexmlgraphicsbatikext [label="batik-ext", shape="box"];
+        |  orgapachexmlgraphicsbatikscript -> orgapachexmlgraphicsbatikext;
+        |  orgapachexmlgraphicsbatikjs [label="batik-js", shape="box"];
+        |  orgapachexmlgraphicsbatikscript -> orgapachexmlgraphicsbatikjs;
+        |  xmlapisxmlapis [label="xml-apis", shape="box"];
+        |  orgapachexmlgraphicsbatikjs -> xmlapisxmlapis;
+        |  xmlapisxmlapis [label="xml-apis", shape="box"];
+        |  orgapachexmlgraphicsbatikscript -> xmlapisxmlapis;
+        |}
+        |""".trimMargin())
+  }
+
   @Test fun singleProjectNoDuplicateDependencyConnections() {
     // Both RxJava and RxAndroid point transitively on reactivestreams.
     singleProject.dependencies.add("implementation", "io.reactivex.rxjava2:rxandroid:2.0.2")

